projects/packages/search/src/instant-search/class-instant-search.php

```diff
@@ -402,6 +408,137 @@ public function auto_config_overlay_sidebar_widgets() {
 		update_option( 'sidebars_widgets', $sidebars );
 	}
 
+	/**
+	 * Add a search widget above footer for block templates.
+	 */
+	public function add_search_block_above_footer() {
+		// We now only checks the core search block.
+		// In the future we need to add Jetpack Search block when it's ready.
+		if ( $this->template_parts_have_search_block() ) {
+			return;
+		}
+
+		$footer = $this->get_template_part( 'footer' );
+		if ( ! $footer instanceof \WP_Block_Template ) {
+			return;
+		}
+
+		$content          = $this->replace_block_patterns( $footer->content );
+		$template_part_id = $footer->id;
+		$request          = new \WP_REST_Request( 'PUT', "/wp/v2/template-parts/{$template_part_id}" );
+		$request->set_header( 'content-type', 'application/json' );
+		$request->set_param( 'content', static::inject_search_widget_to_block( $content ) );
+		$request->set_param( 'id', $template_part_id );
+		$controller = new \WP_REST_Templates_Controller( 'wp_template_part' );
+		return $controller->update_item( $request );
+	}
+
+	/**
+	 * Replace pattern blocks with their content.
+	 * We don't want to replace recursively for the sake of simplicity.
+	 *
+	 * @param string $block_content - Content of template part.
+	 */
+	protected function replace_block_patterns( $block_content ) {
```

Comment:
I don't think I quite understand what this function is doing. What does replacing pattern blocks with their content do for us here?

Reply:
For some templates ( e.g. twentytwentytwo ), the footer  consists of block patterns like:
```
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var(--wp--custom--spacing--large, 8rem)"}}},"layout":{"inherit":true}} -->
<div class="wp-block-group" style="padding-top:var(--wp--custom--spacing--large, 8rem)">
<!-- wp:pattern {"slug":"twentytwentytwo/footer-default"} /-->
</div>
<!-- /wp:group -->

```

The function is to find every block pattern and replace them with the actual content, so that we could find the perfect place to insert the search block. This is also what happen if user edit the block.

Reply:
You could look at the input/output from its test: https://github.com/Automattic/jetpack/pull/22613/files#diff-1e0b7c3a5ac0caab76d82f0b747c1fa627d8ee203f62d53afd375567f3c1be66R229

Reply:
Excellent explanation, thank you!